twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java

```diff
@@ -177,7 +187,22 @@ protected void doStartUp() {
 
     @Override
     protected void doShutDown() {
-      // No-op
+      // Wait for sometime for the container to stop
+      // TODO: Use configurable value for stop time
+      int maxWaitSecs = Constants.APPLICATION_MAX_STOP_SECONDS;
+      try {
+        if (shutdownLatch.await(maxWaitSecs, TimeUnit.SECONDS)) {
+          return;
+        }
+      } catch (InterruptedException e) {
+        LOG.error("Got exception while waiting for runnable {}, instance {} to stop", runnable, instanceId);
+        // TODO: how do we handle the InterruptedException? Should we restore the interrupted status?
+        return;
+      }
+      // Container has not shutdown even after maxWaitSecs after sending stop message,
+      // we'll need to kill the container
+      LOG.warn("Killing runnable {}, instance {} after waiting {} secs", runnable, instanceId, maxWaitSecs);
+      kill();
```

Comment:
In case of exception during a kill, I'll add code to retry the kill a few times.